hbase/src/main/java/magnolify/hbase/HBaseType.scala

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright 2023 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package magnolify.hbase
+
+import magnolia1.{CaseClass, Magnolia, SealedTrait}
+import magnolify.shared.*
+import magnolify.shims.FactoryCompat
+import org.apache.hadoop.hbase.client.{Mutation, Put, Result}
+
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets.UTF_8
+import java.util.UUID
+import scala.annotation.implicitNotFound
+import scala.jdk.CollectionConverters.*
+sealed trait HBaseType[T] extends Converter[T, Map[String, Array[Byte]], Map[String, Array[Byte]]] {
+  def apply(v: Result, family: Array[Byte]): T =
+    from(
+      v.getFamilyMap(family)
+        .asScala
+        .map { case (qualifier, value) => new String(qualifier, UTF_8) -> value }
+        .toMap
+    )
+
+  def apply(v: T, row: Array[Byte], family: Array[Byte], ts: Long = 0L): Mutation = to(v)
+    .foldLeft(new Put(row)) { case (row, (qualifier, value)) =>
+      row.addColumn(family, qualifier.getBytes(UTF_8), ts, value)
+    }
+}
+
+object HBaseType {
+  implicit def apply[T: HBaseField]: HBaseType[T] = HBaseType(CaseMapper.identity)
+
+  def apply[T](cm: CaseMapper)(implicit f: HBaseField[T]): HBaseType[T] = f match {
+    case r: HBaseField.Record[_] =>
+      new HBaseType[T] {
+        private val caseMapper: CaseMapper = cm
+
+        override def from(xs: Map[String, Array[Byte]]): T = r.get(xs, null)(caseMapper).get
+
+        override def to(v: T): Map[String, Array[Byte]] = r.put(null, v)(caseMapper)
+      }
+    case _ =>
+      throw new IllegalArgumentException(s"BigtableType can only be created from Record. Got $f")
+  }
+}
+
+sealed trait HBaseField[T] extends Serializable {
+  def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T]
+  def put(k: String, v: T)(cm: CaseMapper): Map[String, Array[Byte]]
+}
+
+object HBaseField {
+  sealed trait Record[T] extends HBaseField[T]
+
+  sealed trait Primitive[T] extends HBaseField[T] {
+    def size: Option[Int]
+    def fromBytes(v: Array[Byte]): T
+    def toBytes(v: T): Array[Byte]
+
+    override def get(xs: Map[String, Array[Byte]], k: String)(cm: CaseMapper): Value[T] =
+      xs.get(k) match {
+        case Some(v) => Value.Some(fromBytes(v))
+        case None    => Value.None
+      }
+
+    override def put(k: String, v: T)(cm: CaseMapper): Map[String, Array[Byte]] =
+      Map(k -> toBytes(v))
+  }
+
+  // ////////////////////////////////////////////////
```

Comment:
missing a comment here or is it just intended as a linebreak?

Reply:
linebreak as in the other modules